src/data-model-providers/codegen/CodegenDataModelProvider.cpp

```diff
@@ -244,6 +267,43 @@ CHIP_ERROR CodegenDataModelProvider::ServerClusters(EndpointId endpointId,
     VerifyOrReturnValue(endpoint->clusterCount > 0, CHIP_NO_ERROR);
     VerifyOrReturnValue(endpoint->cluster != nullptr, CHIP_NO_ERROR);
 
+    // We have 2 managed lists:
+    //   - ember clusters, that have ember metadata AND ember version
+    //   - `ServerClusterInterfaceRegistry` clusters which MAY have an ember version
+    //     however may also not have one (we should accept server clusters registered
+    //     completely outside ember).
```

Comment:
Why should we accept "server clusters registered outside ember"?  I thought the whole point was to allow having a provider _not_ tied to codegen that things that codegen does not know about are registered with?  Why are we shoehorning totally-non-codegen things into the codegen provider?

Reply:
Need to reword it however unsure how. The key part is "ember version" or generally "cluster version":

- new clusters have `cluster->GetDataVersion()`
- ember will do a `ServerClusterEntryFrom(...)` which calls `emberAfDataVersionStorage` to figure out the version for the cluster

Since they are not the same, I loop through registry clusters first to say their "GetDataVersion" is the authority and then add ember items next.

Reply:
OK, so is the key part this:

> To allow cluster implementations to be incrementally converted to storing their own data versions, instead of relying on the out-of-band emberAfDataVersionStorage, first check for clusters that are using the new data version storage and are registered via ServerClusterInterfaceRegistry, then fill in the data versions for the rest via the out-of-band mechanism.

?

Reply:
Updated the comment. Included your text and a bit more at the top.